### PR TITLE
feat: create backend routes and API for importing saved queries

### PR DESCRIPTION
**Pull request by AAfghahi**
### SUMMARY

Created backend routes and validation for Import Saved Queries feature that Superset is rolling out. 

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
none yet

### TEST PLAN
none yet! 

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset/queries/saved_queries/api.py [hunk 1]
```diff
@@ -252,3 +265,135 @@ def export(self, **kwargs: Any) -> Response:
             as_attachment=True,
             attachment_filename=filename,
         )
+    @expose("/import/", methods=["POST"])
```
Comment (betodealmeida) [inline superset/queries/saved_queries/api.py:268]: You also need to add `RouteMethod.IMPORT` to `include_route_methods` in this file, otherwise the endpoint will not be exposed.

File: superset/queries/saved_queries/api.py [hunk 2]
```diff
@@ -252,3 +265,135 @@ def export(self, **kwargs: Any) -> Response:
...
+    @protect()
+    @safe
+    @statsd_metrics
+    @event_logger.log_this_with_context(
+        action=lambda self, *args, **kwargs: f"{self.__class__.__name__}.import_",
+        log_to_statsd=False,
+    )
+    def import_(self) -> Response:
+        """Import chart(s) with associated datasets and databases
```
Comment (betodealmeida) [inline superset/queries/saved_queries/api.py:277]: chart -> saved_query (same thing in other places)

Comment (AAfghahi) [reply superset/queries/saved_queries/api.py:277]: embarrassing, thank you for pointing it out.

File: superset/queries/saved_queries/api.py [hunk 3]
```diff
@@ -252,3 +265,135 @@ def export(self, **kwargs: Any) -> Response:
...
+            return self.response_500(message=str(exc))
+    @expose("/", methods=["POST"])
+    @protect()
+    @safe
+    @statsd_metrics
+    @event_logger.log_this_with_context(
+        action=lambda self, *args, **kwargs: f"{self.__class__.__name__}.post",
+        log_to_statsd=False,
+    )
+    def post(self) -> Response:
```
Comment (betodealmeida) [inline superset/queries/saved_queries/api.py:349]: We don't need this for this PR.

Creating a new saved query would be an API for when the user clicks "save query" in SQL Lab. Currently that flow uses the old API (`savedqueryviewapi/api/create`), but when we eventually migrate it to the new v1 API this is the endpoint that would be used.

I would suggest moving this out of the PR, and have a separate PR introducing this and the command, so we can move "save query" to this endpoint.

File: superset/queries/saved_queries/commands/create.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```
Comment (betodealmeida) [inline superset/queries/saved_queries/commands/create.py:1]: This is also not needed, let's move to a separate PR.

File: superset/queries/saved_queries/commands/importers/v1/_init_.py [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    import_error = SavedQueryImportError
+
+    @staticmethod
+    def _import(
+        session: Session, configs: Dict[str, Any], overwrite: bool = False
+    ) -> None:
+        # discover datasets associated with saved queries
+        dataset_uuids: Set[str] = set()
+        for file_name, config in configs.items():
+            if file_name.startswith("saved_queries/"):
```
Comment (betodealmeida) [inline superset/queries/saved_queries/commands/importers/v1/_init_.py:52]: Same here, currently the directory is `queries/`. (Check the ZIP file with an exported saved query.)

(We can't change it to `saved_queries/` unless we bump the version of the importer.)

File: superset/queries/saved_queries/commands/importers/v1/_init_.py [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+        "datasets/": ImportV1DatasetSchema(),
+        "saved_query": ImportV1SavedQuerySchema(),
+    }
+    import_error = SavedQueryImportError
+
+    @staticmethod
+    def _import(
+        session: Session, configs: Dict[str, Any], overwrite: bool = False
+    ) -> None:
+        # discover datasets associated with saved queries
```
Comment (betodealmeida) [inline superset/queries/saved_queries/commands/importers/v1/_init_.py:49]: Saved queries are not associated with datasets, they're associated with databases only. This is how the exported saved query looks like:

```
schema: master
label: Untitled Query 2
description: test
sql: SELECT * FROM covid_dashboard WHERE WHERE FROM HA AVG SUM MAX
uuid: 5a8be71f-f59f-4a7c-a7b3-e59394168062
version: 1.0.0
database_uuid: 45812a92-6844-4578-8692-0cc603c3fed0
```

**Comment by betodealmeida** [comment]
Don't forget to add unit tests, you can find examples for all the other imports inside `tests`.

File: superset/queries/saved_queries/api.py (hunk 1 above)
Comment (AAfghahi) [reply superset/queries/saved_queries/api.py:268]: oh, I hadn't noticed that. Added.

File: superset/queries/saved_queries/api.py (hunk 2 above)
Comment (betodealmeida) [reply superset/queries/saved_queries/api.py:277]: I've done that so many times! 😆

File: superset/queries/saved_queries/commands/importers/v1/_init_.py [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+
+class ImportSavedQueriesCommand(ImportModelsCommand):
+    """Import Saved Queries"""
+
+    dao = SavedQueryDAO
+    model_name= "saved_queries"
+    prefix ="saved_queries/"
+    schemas: Dict[str, Schema] = {
+        "datasets/": ImportV1DatasetSchema(),
+        "saved_query": ImportV1SavedQuerySchema(),
```
[the remark this answers is not in the document]
Comment (AAfghahi) [reply superset/queries/saved_queries/commands/importers/v1/_init_.py:41]: by this do you mean changing it to saved_queries or replacing it with queries/ altogether. Furthermore, would I then use ImportV1SavedQuerySchema still?

[the remark this answers is not in the document]
Comment (betodealmeida) [reply superset/queries/saved_queries/commands/importers/v1/_init_.py]: This is used when validating the schema of all the files inside the ZIP: https://github.com/apache/superset/blob/master/superset/commands/importers/v1/__init__.py#L108-L111

If you open the ZIP file from a saved query export you'll see that saved queries are stored inside the `queries/` subdirectory, so we need to use that:

```python
schemas: Dict[str, Schema] = {
        "datasets/": ImportV1DatasetSchema(),
        "queries/": ImportV1SavedQuerySchema(),
```

File: superset/queries/saved_queries/commands/exceptions.py [hunk 8]
```diff
@@ -25,3 +30,10 @@ class SavedQueryBulkDeleteFailedError(DeleteFailedError):
 
 class SavedQueryNotFoundError(CommandException):
     message = _("Saved query not found.")
+
+class SavedQueryImportError(ImportFailedError):
+    message = _("Import chart failed for an unknown reason")
```
Comment (betodealmeida) [inline superset/queries/saved_queries/commands/exceptions.py:35]: End the message in a period for consistency.

File: superset/queries/saved_queries/commands/exceptions.py [hunk 9]
```diff
@@ -25,3 +30,10 @@ class SavedQueryBulkDeleteFailedError(DeleteFailedError):
...
+
+class SavedQueryInvalidError(CommandInvalidError):
+    message = _("Saved Query Parameters are invalid")
```
Comment (betodealmeida) [inline superset/queries/saved_queries/commands/exceptions.py:38]:
```
message = _("Saved Query parameters are invalid.")
```

File: superset/queries/saved_queries/commands/importers/v1/_init_.py [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+from superset.queries.dao import SavedQueryDAO
+from superset.queries.saved_queries.commands.importers.v1.utils import import_saved_query
+from superset.queries.saved_queries.schemas import ImportV1SavedQuerySchema
+
+class ImportSavedQueriesCommand(ImportModelsCommand):
+    """Import Saved Queries"""
+
+    dao = SavedQueryDAO
+    model_name= "saved_queries"
+    prefix ="saved_queries/"
```
Comment (betodealmeida) [inline superset/queries/saved_queries/commands/importers/v1/_init_.py:38]: This also needs to be `queries/`, matching the name of the directory inside the ZIP file.

File: superset/queries/saved_queries/commands/importers/v1/_init_.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (AAfghahi) [reply superset/queries/saved_queries/commands/importers/v1/_init_.py]: ok that makes a lot of sense. Thank you.

File: superset/queries/saved_queries/commands/importers/v1/_init_.py (hunk 6 above)
Comment (AAfghahi) [reply superset/queries/saved_queries/commands/importers/v1/_init_.py:49]: ` def _import(
        session: Session, configs: Dict[str, Any], overwrite: bool = False
    ) -> None:
        # discover databases associated with saved queries
        dataset_uuids: Set[str] = set()
        for file_name, config in configs.items():
            if file_name.startswith("queries/"):
                dataset_uuids.add(config["database_uuid"])

        # import related databases
        database_ids: Dict[str, int] = {}
        for file_name, config in configs.items():
            if file_name.startswith("databases/") and config["uuid"] in database_uuids:
                database = import_database(session, config, overwrite=False)
                database_ids[str(database.uuid)] = database.id


        # import saved queries with the correct parent ref
        for file_name, config in configs.items():
            if file_name.startswith("queries/") and config["database_uuid"] in database:
                # update datasource id, type, and name
                database = database[config["dataset_uuid"]]
                config.update(
                    {
                        "datasource_id": database.id,
                        "datasource_name": database.table_name,
                    }
                )
                config["params"].update({"datasource": database.uid})
                import_saved_query(session, config, overwrite=overwrite)`

File: superset/queries/saved_queries/commands/importers/v1/__init__.py [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+        database_ids: Dict[str, int] = {}
+        for file_name, config in configs.items():
+            if file_name.startswith("databases/") and config["uuid"] in database_uuids:
+                database = import_database(session, config, overwrite=False)
+                database_ids[str(database.uuid)] = database.id
+
+
+        # import saved queries with the correct parent ref
+        for file_name, config in configs.items():
+            if file_name.startswith("queries/") and config["database_uuid"] in database:
```
Comment (betodealmeida) [inline superset/queries/saved_queries/commands/importers/v1/__init__.py:65]: `database` here is a `Database` object, you want:

```suggestion
            if file_name.startswith("queries/") and config["database_uuid"] in database_ids:
```

File: superset/queries/saved_queries/commands/importers/v1/__init__.py [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+                # update datasource id, type, and name
+                database = database[config["dataset_uuid"]]
+                config.update(
+                    {
+                        "datasource_id": database.id,
+                        "datasource_name": database.table_name,
+                    }
+                )
+                config["params"].update({"datasource": database.uid})
```
Comment (betodealmeida) [inline superset/queries/saved_queries/commands/importers/v1/__init__.py:74]: A saved query has a many-to-one relationship to database; there are no datasources involved. What you need to do here is update the [`db_id` attribute of the saved query model](https://github.com/apache/superset/blob/master/superset/models/sql_lab.py#L179) and point it to the imported/existing database:

```suggestion
                config["db_id"] = database_ids[config["database_uuid"]]
```

File: superset/queries/saved_queries/api.py [hunk 13]
```diff
@@ -252,3 +262,76 @@ def export(self, **kwargs: Any) -> Response:
...
+    @expose("/import/", methods=["POST"])
+    @protect()
+    @safe
+    @statsd_metrics
+    @event_logger.log_this_with_context(
+        action=lambda self, *args, **kwargs: f"{self.__class__.__name__}.import_",
+        log_to_statsd=False,
+    )
+    def import_(self) -> Response:
+        """Import Saved Queries with associated datasets and databases
```
Comment (betodealmeida) [inline superset/queries/saved_queries/api.py:274]:
```suggestion
        """Import Saved Queries with associated databases
```

File: superset/queries/saved_queries/api.py [hunk 14]
```diff
@@ -252,3 +262,76 @@ def export(self, **kwargs: Any) -> Response:
...
+                  properties:
+                    formData:
+                      description: upload file (ZIP)
+                      type: string
+                      format: binary
+                    passwords:
+                      description: JSON map of passwords for each file
+                      type: string
+                    overwrite:
+                      description: overwrite existing databases?
```
Comment (betodealmeida) [inline superset/queries/saved_queries/api.py:292]:
```suggestion
                      description: overwrite existing saved queries?
```

File: superset/queries/saved_queries/commands/importers/v1/__init__.py [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+from superset.datasets.schemas import ImportV1DatasetSchema
+from superset.queries.saved_queries.dao import SavedQueryDAO
+from superset.queries.saved_queries.commands.importers.v1.utils import import_saved_query
+from superset.queries.saved_queries.schemas import ImportV1SavedQuerySchema
+
+class ImportSavedQueriesCommand(ImportModelsCommand):
+    """Import Saved Queries"""
+
+    dao = SavedQueryDAO
+    model_name= "saved_queries"
```
Comment (betodealmeida) [inline superset/queries/saved_queries/commands/importers/v1/__init__.py:37]: The lint is going to fail here and in a few other places. You can auto format it by running `pre-commit run --all-files`.

```suggestion
    model_name = "saved query"
```

File: superset/queries/saved_queries/commands/importers/v1/__init__.py [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+    prefix ="saved_queries/"
```
Comment (betodealmeida) [inline superset/queries/saved_queries/commands/importers/v1/__init__.py:38]: This should be the subdirectory inside the ZIP where the relevant YAML files exist:

```suggestion
    prefix = "queries/"
```

File: tests/charts/commands_tests.py [hunk 17]
```diff
@@ -197,7 +197,7 @@ def test_import_v1_chart(self):
         db.session.commit()
 
     def test_import_v1_chart_multiple(self):
-        """Test that a dataset can be imported multiple times"""
+        """Test that a chart can be imported multiple times"""
```
Comment (betodealmeida) [inline tests/charts/commands_tests.py:200]: Thanks! :)

File: tests/queries/saved_queries/api_tests.py [hunk 18]
```diff
@@ -745,3 +751,48 @@ def test_export_not_allowed(self):
...
+            with bundle.open(
+                "saved_query_export/databases/imported_database.yaml", "w"
+            ) as fp:
+                fp.write(yaml.safe_dump(database_config).encode())
+            with bundle.open("saved_query_export/queries/imported_database/public/imported_saved_query.yaml", "w") as fp:
+                fp.write(yaml.safe_dump(saved_queries_config).encode())
+        buf.seek(0)
+        return buf
+
+    @pytest.mark.usefixtures("create_saved_queries")
```
Comment (betodealmeida) [inline tests/queries/saved_queries/api_tests.py:769]: A [fixture](https://docs.pytest.org/en/stable/fixture.html) runs setup code before a test function, and cleanup code after. There's no fixture called `create_saved_queries`, so you should remove this.

File: tests/queries/saved_queries/commands_tests.py [hunk 19]
```diff
@@ -108,3 +120,109 @@ def test_export_query_command_key_order(self, mock_g):
...
+            """
+            -- Note: Unless you save your query, 
+            these tabs will NOT persist if you clear
+            your cookies or change browsers.
+
+            SELECT * from birth_names
+            """
+        )
+        assert saved_query.uuid == "05b679b5-8eaf-452c-b874-a7a774cfa4e9"
+        assert saved_query.database_uuid == "b8a1ccd3-779d-4ab7-8ad8-9ab119d7fe89"
```
Comment (betodealmeida) [inline tests/queries/saved_queries/commands_tests.py:149]: I suspect these will change every time you run the tests, so you might need to remove these tests.

File: tests/queries/saved_queries/commands_tests.py [hunk 20]
```diff
@@ -108,3 +120,109 @@ def test_export_query_command_key_order(self, mock_g):
...
+        assert saved_query.schema == "public"
+        assert saved_query.sql == (
+            """
+            -- Note: Unless you save your query, 
+            these tabs will NOT persist if you clear
+            your cookies or change browsers.
+
+            SELECT * from birth_names
+            """
+        )
```
Comment (betodealmeida) [inline tests/queries/saved_queries/commands_tests.py:147]: This test will probably fail because the SQL is formatted differently (there's no line break after "your query,", eg).

File: tests/queries/saved_queries/commands_tests.py [hunk 21]
```diff
@@ -108,3 +120,109 @@ def test_export_query_command_key_order(self, mock_g):
...
+        saved_query = db.session.query(SavedQuery).filter_by(datasource_id=database.id).all()
+        assert len(saved_query) == 1
+
+        db.session.delete(saved_query[0])
+        db.session.delete(database)
+        db.session.commit()
+        
+    def test_import_v1_saved_queries_validation(self):
+        """Test different validations applied when importing a chart"""
```
Comment (betodealmeida) [inline tests/queries/saved_queries/commands_tests.py:179]:
```suggestion
        """Test different validations applied when importing a saved query"""
```

File: tests/queries/saved_queries/commands_tests.py [hunk 22]
```diff
@@ -108,3 +120,109 @@ def test_export_query_command_key_order(self, mock_g):
...
+        # metadata.yaml must be present
+        contents = {
+            "metadata.yaml": yaml.safe_dump(saved_queries_metadata_config),
```
Comment (betodealmeida) [inline tests/queries/saved_queries/commands_tests.py:182]: You need to remove `metadata.yaml`, since here you're testing that a validation message "Missing metadata.yaml" is shown when the file is missing.

**Review by betodealmeida — CHANGES_REQUESTED** [review]

File: superset/queries/saved_queries/commands/importers/v1/__init__.py (hunk 12 above)
Comment (AAfghahi) [reply superset/queries/saved_queries/commands/importers/v1/__init__.py:74]: That makes sense, thank you for showing me the source of what I am supposed to be looking at, I appreciate it. Quick clarifying question: `config["db_id"] = database_ids[config["database_uuid"]]` should come right after config["params"] (on line 74). right? I think this is a weird github thing because it shows the line after params, but the line number is 66.

File: superset/queries/saved_queries/commands/importers/v1/__init__.py (hunk 15 above)
Comment (AAfghahi) [reply superset/queries/saved_queries/commands/importers/v1/__init__.py:37]: ok that saves me a lot of heart ache. ty

File: tests/queries/saved_queries/api_tests.py (hunk 18 above)
Comment (AAfghahi) [reply tests/queries/saved_queries/api_tests.py:769]: that makes sense, sort of like a beforeEach afterEach in RTL. Would it be more thorough to add this fixture?

File: tests/queries/saved_queries/commands_tests.py (hunk 20 above)
Comment (AAfghahi) [reply tests/queries/saved_queries/commands_tests.py:147]: Yeah, they do. Was wondering about this actually, should I just not test the sql or is there a comment to disable the line is too long linter?

File: superset/queries/saved_queries/commands/importers/v1/__init__.py (hunk 15 above)
Comment (AAfghahi) [reply superset/queries/saved_queries/commands/importers/v1/__init__.py:37]: wait, so I set up a virtual env and run: 

`superset pre-commit run --all-files`

when I do that it tells me that there is no pre-commit command. Or do I do the pre-commit as part of my git statement?

**Review by AAfghahi — COMMENTED** [review]
Thank you

**Review by AAfghahi — COMMENTED** [review]
ty

File: superset/queries/saved_queries/commands/importers/v1/__init__.py (hunk 12 above)
Comment (betodealmeida) [reply superset/queries/saved_queries/commands/importers/v1/__init__.py:79]: The change I suggested deletes lines 71-79, since datasources are not used in saved queries and replaces them with `config["db_id"] = database_ids[config["database_uuid"]]`.

You should be able to click "Commit suggestion" to automatically apply the suggested change, but then you need to pull the changes to your local branch if you need to modify it again. Or is the button not there for you?

File: tests/queries/saved_queries/commands_tests.py (hunk 20 above)
Comment (betodealmeida) [reply tests/queries/saved_queries/commands_tests.py]: You can disable to line it too linter, or break the line:

```python
assert saved_query == (
    "-- Note: Unless you save your query, "
    "these tabs will NOT persist if you clear "
    ...
)

**Review by betodealmeida — APPROVED** [review]

**Comment by AAfghahi** [comment]
@betodealmeida Thank you! Though, its still failing two tests :/

File: superset/queries/saved_queries/commands/importers/v1/utils.py [hunk 23]
```diff
@@ -0,0 +1,38 @@
...
+from sqlalchemy.orm import Session
+
+from superset.models.sql_lab import SavedQuery
+
+
+def import_saved_query(
+    session: Session, config: Dict[str, Any], overwrite: bool = False
+) -> SavedQuery:
+    existing = session.query(SavedQuery).filter_by(uuid=config["uuid"]).first()
+    if existing:
```
Comment (hughhhh) [inline superset/queries/saved_queries/commands/importers/v1/utils.py:29]: nit:
```suggestion
if existing and not overwrite:
          return existing
```

File: superset/queries/saved_queries/commands/importers/v1/utils.py [hunk 24]
```diff
@@ -0,0 +1,38 @@
...
+        if not overwrite:
+            return existing
+        config["id"] = existing.id
```
Comment (hughhhh) [inline superset/queries/saved_queries/commands/importers/v1/utils.py:32]: nit:
```suggestion
    if existing and not overwrite:
              return existing
    config["id"] = existing.id
```

**Review by hughhhh — APPROVED** [review]
